### ISSUE TITLE
oDataProvider handling of DELETE entity requests fails

### ISSUE THREAD

**Issue by shturec** [jaystack/jaydata#110]
When a request for DELETE is sent (entity.remove()), the response handler in oDataProvider.js processes the response data as it does with GET requests since the HTTP response codes for both are 200 (as defined by the spec) and that's the only condition taken into account (see the response handler function within the _saveRest method on line 668). 
However, as defined by the spec as well, DELETE requests do not return response body in contrast to the query requests. The current implementation does not account for that, and assumes that the data parameter is never undefined (which is the case with DELETE responses as defined in the spec). And so the next function invocation - reload_fromResponse fails as expected.
I did a quick and dirty hack that checks for DELETE requests when the response code is 200 and continues to the reload function only if it's not. This prevented trying to load from an undefined data (response body) in case of DELETEs. It worked fine for me but I suppose that an eye that's more careful and experienced with jaydata and the oDataProvider is needed. 
Jaydata version is 1.3.1

**Comment by agdolla** [comment]
Hi
Thanks for reporting this problem, could you please share your quick and dirty hack with us ?
Btw, the specification is vague, in the html spec it says that the server must return '200 OK' but in the pdf version '204 No Content'.

**Comment by shturec** [comment]
I haven't seen the PDF version but it's too bad for a specification to mess up like this :-( I didn't get the intention behind return code 200 with no body anyways and this is what unfortunately odata4j chose to return. It might have been a typo or a leftover as well, because as I see in v3 this has been already changed to return code 204 for both removing entities and links:
http://www.odata.org/documentation/odata-v3-documentation/odata-core/#1034_Delete_an_Entity 

Despite all this, I believe that a small check if there's a body to process before doing so and if the method is DELETE so that it's considered a valid case for legacy (v2) applications would not hurt but will make things smoother for us all.

I couldn't find a way to pass the request object to the response handler and use it to check if the request method has been DELETE. So i raise a global flag during the request processing phase. Inside the response handler I check for the flag and if it's there it is deleted and the data is not processed. Like I said - quick and dirty :-)

```
if($data.EntityState.Deleted){
    window['isDeleteRequest'] = true;
}

var requestData = [request, function (data, response) {
    if (response.statusCode >= 200 && response.statusCode < 300) {
        var item = convertedItem[0];

        if (response.statusCode == 204) {
            if (response.headers.ETag || response.headers.Etag || response.headers.etag) {
                var property = item.getType().memberDefinitions.getPublicMappedProperties().filter(function (memDef) { return memDef.concurrencyMode === $data.ConcurrencyMode.Fixed });
                if (property && property[0]) {
                    item[property[0].name] = response.headers.ETag || response.headers.Etag || response.headers.etag;
                }
            }
        } else {
            if(window['isDeleteRequest']){
                delete window['isDeleteRequest'];
            } else {
                that.reload_fromResponse(item, data, response);
            }
        }

        if (callBack.success) {
            callBack.success(convertedItem.length);
        }
    } else {
        callBack.error(that.parseError(response));
    }

}, function (e) {
    callBack.error(that.parseError(e));
}];
```

**Comment by shturec** [comment]
By the way, I need to do exactly the same to overcome yet another deficiency in odata4j, this time wrt the return code for MERGE operations. They return 200, while oDataProvider would be happier with 204. Therefore in _saveRest in oDataProvider.js before the setup of the requestData variable:

```
var isMergeRequest;
if(request.headers['X-HTTP-METHOD'] && 'MERGE'===request.headers['X-HTTP-METHOD']){
   window['isMergeRequest'] = true;
}
```

and then in the response handler

```
if(window['isMergeRequest']){
    response.statusCode = 204;
    delete window['isMergeRequest'];
}
```

Or even better - use the 'that' variable. However either of these does not look as well as having a reference to the request in the response handler or something like this.
Of course, this is not a bug on your side but a fix for the originally reported issue might lead to a better solution to this problem as well. Even if it requires tampering with jaydata itself again.

**Comment by agdolla** [comment]
Sorry, too complicated :)
We'll solve this issue, one easy patch will be to handle the situation, when (statusCode == 200 and response.headers['Content-Length'] == 0) then do do parse it at all, just like in the case of 204

**Comment by shturec** [comment]
Yes, that sounds sensible for the DELETE and would probably work no matter how compliant the odata provider is with the specification.
Is there a roadmap for the patch?
Thanks!
